### PR TITLE
Implementing job rescheduling for the queued job

### PR DESCRIPTION
**Pull request by ScopeyNZ**
Was previously #93

> This PR makes the job that checks for installed version/updates reschedule itself after it completes. As expected, this conflicted with the way the refresh button determined whether it should display as clickable or in the "updating" state. Here's how it works after this PR:
> 
> - Currently when the build script is run the job gets queued in the normal "queued" queue
> - When the job completes, it now adds itself back on the queue but indicates it should only start after 24 hours.
> - The refresh button now only looks at the "immediate" queue for jobs
> - If the refresh button is clicked, it will try to add the job to the immediate queue.
> - Given the way the queued job service works, if the job is currently queued, it will do nothing but return the ID of the queued job
> - The job descriptor is fetched, and if it is not on the immediate queue we move it to the immediate queue and ensure that it's set to start immediately
> - When the job finishes it will reschedule for 24 hours again
> 
> This means that after any run (even pressing the refresh button) the next unprompted run of the job will be 24 hours after the last run.
> 
> Do you think I should add config settings for both the delay time and whether scheduling should be done at all?

### CODE REVIEW

File: src/Forms/GridFieldRefreshButton.php [hunk 1]
```diff
@@ -164,8 +182,37 @@ public function hasActiveJob()
     public function handleRefresh()
     {
         if (!$this->hasActiveJob()) {
-            $injector = Injector::inst();
-            $injector->get(QueuedJobService::class)->queueJob($injector->create(CheckForUpdatesJob::class));
+            // Queue the job in the immediate queue
```
Comment (robbieaverill) [inline src/Forms/GridFieldRefreshButton.php]: `if ($this->hasActiveJob()) { return; }` might look slightly cleaner - thoughts?

File: src/Forms/GridFieldRefreshButton.php [hunk 2]
```diff
@@ -164,8 +182,37 @@ public function hasActiveJob()
...
+    public function getQueuedJobService()
+    {
+        return $this->queuedJobService;
+    }
+
+    /**
+     * @param QueuedJobService $queuedJobService
+     * @return $this
+     */
+    public function setQueuedJobService($queuedJobService)
```
Comment (robbieaverill) [inline src/Forms/GridFieldRefreshButton.php]: Can you please type hint QueuedJobService

File: src/Jobs/CheckForUpdatesJob.php [hunk 3]
```diff
@@ -13,10 +13,7 @@ class CheckForUpdatesJob extends AbstractQueuedJob implements QueuedJob
      */
     public function getTitle()
     {
-        return _t(
-            'CheckForUpdates.TITLE',
-            'Check for updates to installed modules'
-        );
```
Comment (robbieaverill) [inline src/Jobs/CheckForUpdatesJob.php]: I'm guessing this wasn't deliberate =)

File: src/Forms/GridFieldRefreshButton.php (hunk 1 above)
Comment (ScopeyNZ) [reply src/Forms/GridFieldRefreshButton.php]: 👍Agreed. Return early

File: src/Jobs/CheckForUpdatesJob.php [hunk 4]
```diff
@@ -48,4 +45,21 @@ public function process()
...
+     * @inheritdoc
+     */
+    public function afterComplete()
+    {
+        // Queue a new job to run in the future
+        $injector = Injector::inst();
+        $queuedJobService = $injector->get(QueuedJobService::class);
+
+        $startAfter = new DateTime(SS_Datetime::now()->getValue());
+        $startAfter->modify('+1 day');
```
Comment (robbieaverill) [inline src/Jobs/CheckForUpdatesJob.php]: Can you make the time period a config string (default: `+1 day`)? Also whether to enable this functionality or not should be configurable (default: true)

File: src/Jobs/CheckForUpdatesJob.php (hunk 3 above)
Comment (ScopeyNZ) [reply src/Jobs/CheckForUpdatesJob.php]: 😅

File: src/Forms/GridFieldRefreshButton.php (hunk 2 above)
Comment (ScopeyNZ) [reply src/Forms/GridFieldRefreshButton.php]: 👍

File: src/Jobs/CheckForUpdatesJob.php (hunk 4 above)
Comment (ScopeyNZ) [reply src/Jobs/CheckForUpdatesJob.php]: I've added configuration options now.

File: src/Jobs/CheckForUpdatesJob.php [hunk 5]
```diff
@@ -48,4 +64,29 @@ public function process()
...
+        if ($reschedule === false) {
+            return;
+        }
+
+        // Queue a new job to run in the future
+        $injector = Injector::inst();
+        $queuedJobService = $injector->get(QueuedJobService::class);
+
+        $startAfter = new DateTime(SS_Datetime::now()->getValue());
+        $startAfter->modify('+' . $rescheduleDelay);
```
Comment (robbieaverill) [inline src/Jobs/CheckForUpdatesJob.php]: Bug: you have `+` twice now - once in the config value and once here

Comment (ScopeyNZ) [reply src/Jobs/CheckForUpdatesJob.php]: I did notice that I forgot to fix it :(

**Comment by robbieaverill** [comment]
Merge on green!

**Pull request merged**